### ISSUE TITLE
Global Store and Vue3 Discussion

### ISSUE THREAD

**Issue by Mootook** [davestewart/vue-class-store#12]
Just replying to https://github.com/davestewart/vue-class-store/issues/11#issuecomment-726939231
As I'd like to discuss more about vue-class-store.

I agree about computed properties being a nice workflow, and it's what I've grown used to, but I'm wondering your thoughts on how vue-class-store can work with the composition api.

From my understanding this package is best used as a provider at the top level then subsequently injected where needed.
I've setup a [codesandbox](https://codesandbox.io/s/vue-store-example-0ksyc?file=/src/App.vue) to try and work through some possible setups. I assume much of the same principles carry along with modifying the store.

Please let me know if you have thoughts on best ways to inject specific store properties at the child level. 
Thanks!

**Comment by davestewart** [comment]
For sure!

So, caveat: I've not done that much with the Composition API as yet.

Just checking your sandbox and the API docs, so that's the Composition API's provide in action, huh?

Good to see it just working!

The demos have example of a global inject in each of them, though that is using the Options API for Vue 3:

- https://github.com/davestewart/vue-class-store-demos

There is a working sandbox here:

- https://codesandbox.io/s/github/davestewart/vue-class-store-demos/tree/master/nuxt?file=/pages/examples/global.vue

As mentioned I'm working on some more global stuff as soon as I get time, but it's more Vue 2 based – as that's what all my projects are right now – but I'll try to get some more Vue 3 under my belt, and feel free to keep dropping your findings here :)

**Comment by djmaze** [comment]
@Mootook Really cool and obvious approach by using `useStore` for injection. I like it.

As we have multiple stores in our app (and we also use typescript), I built a generic wrapper module for injecting stores. You can find it in [this sandbox](https://codesandbox.io/s/vue-store-example-with-generic-provideinject-rkgy0?file=/src/components/HelloWorld.vue). It would be cool to be able to get rid of the separate `useStore()` and `injectStore()` though somehow.

**Comment by Mootook** [comment]
Curious about potential overhead with injecting across the app.
From the docs:

> provide and inject are primarily provided for advanced plugin / component library use cases. It is NOT recommended to use them in generic application code.

Anyone have any insight on the comparison between directly injecting the store vs passing it via props? Injecting seems fine, recommended even, for grandchild components, but since this is functioning like a pseudo-global store, wondering if the concepts still hold for every level of the component hierarchy.

**Comment by davestewart** [comment]
> provide and inject are primarily provided for advanced plugin / component library use cases. It is NOT recommended to use them in generic application code.

This is increasingly taken to be fairly Vue 2-opinionated, and perhaps not even accurate for Vue 2.

My impression was at the time, that was what the feature was intended for, and the developers were more keen to shepherd people towards Vuex, but the last few years have been a learning experience, and provide/inject now compares favourably in terms of outcome/overhead.

TL;DR not so relevant now

**Comment by davestewart** [comment]
FWIW Element UI does a lot of injecting for its form elements. Some of it is really rather clever:

- https://github.com/ElemeFE/element/search?q=inject

**Comment by Mootook** [comment]
Ah, okay. Makes sense. Thank you.

**Comment by Mootook** [comment]
As for the global state, what are your thoughts on creating a static property on the root of the store which can be accessed by child classes? I'm mainly looking for a way to communicate across nested objects/classes within the store and not necessarily have to chain things together in some component...codesandbox has some issues disallowing eslint which makes a demo difficult but here's the rough idea:
``` javascript
class Store {
  static instance

  constructor () {
    this.nestedClass = new NestedClass()
    this.instance = this
  }

  bar () {
    console.log('Called from a nested class')
  }
}

class NestedClass {
  foo () {
    Store.instance.bar() // Called from a nested class
  }
}

const store = new Store()
store.nestedClass.foo()
```


This small test case "works", I just want to make sure it's not going against what the utility of this library.

**Comment by davestewart** [comment]
Yeah, it's totally up to you.

I've been pondering various ideas, some more `registerStore()` type helpers with a global `state` POJO (as the top level may not need to be reactive).

Or maybe (because the decorator is a factory function) `this.$store` or `this.$root` is injected into each instance.

That would likely need some additional augmentation of the `VueStore` type, which isn't a problem, would just need to talk to people and find out what feels like the best option; maybe it would be best left to users to augment their own stores:

- https://www.digitalocean.com/community/tutorials/typescript-module-augmentation

**Comment by Mootook** [comment]
I don't use TypeScript in my project, and the only issue I had was defining properties outside the constructor. 
Also, `Store.instance = this`, I found to have issues with reactivity, since I'm not sure if it's a vue instance at that point in the constructor.

**Comment by davestewart** [comment]
OK, good info, thanks.

**Comment by davestewart** [comment]
Hi all,

Just checking in on the global store stuff.

It may be too early to declare victory, but I appear to have gotten a nice TypeScript workflow going using:

- a `RootStore` class instance
- a global `Vue.use()`
- and [Module Augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) which provides autocomplete throughout the project

The root store class looks like this:

```ts
export class RootStore {
  $add (name: string, instance) {
    if (name.startsWith('$')) {
      throw new Error('The "$" character is reserved for store instance methods')
    }
    this[name] = instance
  }

  $remove (name: string) {
    delete this[name]
  }

  constructor (modules: Record<string, any> = {}) {
    Object.keys(modules).forEach(key => {
      this.$add(key, modules[key])
    })
  }
}
```

My store index looks like this:

```ts
import Vue from 'vue'
import { RootStore, install } from '@/packages/store'

// local modules
import app, { AppStore } from './app'
import ui, { UiStore } from './ui'

// tell typescript about modules
declare module '@/packages/store' {
  interface RootStore {
    app: AppStore,
    ui: UiStore,
  }
}

// tell vue about the built-in store
declare module 'vue/types/vue' {
  interface Vue {
    $store: RootStore
  }
}

// install and add modules
Vue.use(install, {
  app,
  ui,
})

// ensure other modules importing store directly know what it is
export default Vue.prototype.$store as RootStore
```

You can add modules individually using a combination of instantiation and augmentation:

```ts
// import root store
import store from '@/store'

// augment root store with the new property 
declare module '@/packages/store' {
  interface RootStore {
    tools: ToolsStore,
  }
}

// local vue class store setup
@VueStore
class ToolsStore {
  resize () {
    bus.call('tools/windows/resize')
  }
}

// physically add the local store to the root store
store.$add('tools', new ToolsStore())
```

Autocomplete works great in Vue files:

![image](https://user-images.githubusercontent.com/132681/102559447-ce600180-40c7-11eb-9f36-8968ee9e2000.png)

I know some of the TS boilerplate looks a little funky, but:

- it allows a root store to be created without explicitly declaring properties
- it allows local stores to be added independently of the root store declarations

I have a minor issue with the `RootStore::$add` not being recognised in one file without me adding the signature to the root augmentation. I don't know why this is (but I'm not an expert in this). Once that is done; everything plays ball.

Happy to hear any thoughts!
